File: android/app/src/test/java/com/clingsync/android/UploadProgressIoTest.kt
package com.clingsync.android

import androidx.test.ext.junit.runners.AndroidJUnit4
import com.clingsync.android.data.UploadProgressIo
import com.clingsync.android.data.UploadStatus
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Rule
import org.junit.Test
import org.junit.rules.TemporaryFolder
import org.junit.runner.RunWith
import org.robolectric.annotation.Config

@RunWith(AndroidJUnit4::class)
@Config(sdk = [28])
class UploadProgressIoTest {
    @get:Rule
    val tmp = TemporaryFolder()

    @Test
    fun progressMapsEveryStatus() {
        val file = tmp.newFile("status.json")
        UploadProgressIo.write(
            file,
            mapOf(
                "a" to UploadStatus.WAITING,
                "b" to UploadStatus.UPLOADING,
                "c" to UploadStatus.UPLOADED,
                "d" to UploadStatus.SKIPPED,
                "e" to UploadStatus.COMMITTING,
            ),
        )

        val progress = UploadProgressIo.readProgress(file)

        assertEquals(FileStatus.Waiting, progress["a"])
        assertEquals(FileStatus.Uploading, progress["b"])
        assertEquals(FileStatus.Uploaded, progress["c"])
        assertEquals(FileStatus.Exists(""), progress["d"])
        assertEquals(FileStatus.Committing, progress["e"])
    }

    @Test
    fun resultKeepsTerminalStatusesAndDropsInProgressOnes() {
        val file = tmp.newFile("result.json")
        UploadProgressIo.write(
            file,
            mapOf(
                "committed" to UploadStatus.COMMITTING,
                "uploaded" to UploadStatus.UPLOADED,
                "skipped" to UploadStatus.SKIPPED,
                "waiting" to UploadStatus.WAITING,
                "uploading" to UploadStatus.UPLOADING,
            ),
        )

        val result = UploadProgressIo.readResult(file)

        assertEquals(FileStatus.Done, result["committed"])
        assertEquals(FileStatus.Done, result["uploaded"])
        assertEquals(FileStatus.Exists(""), result["skipped"])
        // Non-terminal states must not overwrite existing UI state on completion.
        assertNull(result["waiting"])
        assertNull(result["uploading"])
        assertEquals(3, result.size)
    }

    @Test
    fun unknownWireStringIsTreatedAsNewAndNonTerminal() {
        val file = tmp.newFile("legacy.json")
        file.writeText("""{"x":"some-future-status"}""")

        assertEquals(FileStatus.New, UploadProgressIo.readProgress(file)["x"])
        assertNull(UploadProgressIo.readResult(file)["x"])
    }
}

File: ios/ClingSyncTests/UploadProgressIoTest.swift
import Foundation
import Testing

@testable import ClingSync

struct UploadProgressIoTest {
    private func tempFile() -> URL {
        FileManager.default.temporaryDirectory.appendingPathComponent("progress-\(UUID().uuidString).json")
    }

    @Test func progressMapsEveryStatus() throws {
        let file = tempFile()
        try UploadProgressIo.write(
            [
                "a": .waiting,
                "b": .uploading,
                "c": .uploaded,
                "d": .skipped,
                "e": .committing,
            ], to: file)

        let progress = try UploadProgressIo.readProgress(from: file)

        #expect(progress["a"] == .waiting)
        #expect(progress["b"] == .sending)
        #expect(progress["c"] == .sentWaitingCommit)
        #expect(progress["d"] == .exists(repoPath: ""))
        #expect(progress["e"] == .committing)
    }

    @Test func resultKeepsTerminalStatusesAndDropsInProgressOnes() throws {
        let file = tempFile()
        try UploadProgressIo.write(
            [
                "committed": .committing,
                "uploaded": .uploaded,
                "skipped": .skipped,
                "waiting": .waiting,
                "uploading": .uploading,
            ], to: file)

        let result = try UploadProgressIo.readResult(from: file)

        #expect(result["committed"] == .done)
        #expect(result["uploaded"] == .done)
        #expect(result["skipped"] == .exists(repoPath: ""))
        // Non-terminal states must not overwrite existing UI state on completion.
        #expect(result["waiting"] == nil)
        #expect(result["uploading"] == nil)
        #expect(result.count == 3)
    }

    @Test func unknownWireStringIsTreatedAsNewAndNonTerminal() throws {
        let file = tempFile()
        try Data(#"{"x":"some-future-status"}"#.utf8).write(to: file)

        #expect(try UploadProgressIo.readProgress(from: file)["x"] == .new)
        #expect(try UploadProgressIo.readResult(from: file)["x"] == nil)
    }
}
